src/jit/gcencode.cpp

```diff
@@ -3549,19 +3550,40 @@ void GCInfo::gcInfoBlockHdrSave(GcInfoEncoder* gcInfoEncoder, unsigned methodSiz
         case TYP_STRUCT:
         {
             CORINFO_CLASS_HANDLE structType = compiler->info.compMethodInfo->args.retTypeClass;
-            if (compiler->IsMultiRegReturnedType(structType) && !compiler->IsHfa(structType))
+            var_types retType = compiler->getReturnTypeForStruct(structType);
+
+            switch (retType)
             {
-                BYTE gcPtrs[2] = {TYPE_GC_NONE, TYPE_GC_NONE};
-                compiler->info.compCompHnd->getClassGClayout(structType, gcPtrs);
+            case TYP_REF:
+            case TYP_ARRAY:
```

Comment:
TYP_ARRAY is never returned by getReturnTypeForStruct, it is confusing to include it here.